rfcs/2020-05-25-more-usable-logevents.md

```diff
@@ -0,0 +1,223 @@
+
+Vector defines success by the highest number of `Event` lifetimes to have been sourced, processed, and then sunk as possible, while maintaining practical levels of reliability and safety.
+
+**If necessary:** Providing a simple, ergonomic API into an internally complex and performant structure is a **worthy trade-off** given the nature of this structure.
+
+For the most part, `LogEvent`s are fairly small, uncomplicated structures, containing a few keys with simple values. Here is a simple example:
+
+```json
+{
+	"message": "I found a potato!",
+	"potato_harvester_id": 123,
+	"potato_location": { "x": 123, "y": 456, }
+}
+
+```
+
+Not so bad, right?
+
+## A Motivatingly Complex Example
+
+Before we can fully grip at our internal `LogEvent` type, let's review a particularly complex example JSON to help us understand where our limits might be:
+
+```json
+// Note, this does not cover all possibilities!
+{
+	"nulled": null, // Ensure a nulled field is not lost.
+	"basic": true,  // Maps can contain multiple types.
+	"list": [
+		true, // Lists can contain multiple types.
+		null, // A null in a list is totally valid.
+		[true, null, true],
+		{
+			"basic": true,
+			"buddy": 1.0, // Maps with multiple values are a bit more complex.
+		}.
+	],
+	"map": {
+		"basic": true,
+		"list": [true, null, true],
+		"map": {
+			"basic": true,
+			"buddy": -1,
+		}
+	},
+}
+
+```
+
+## Dogma and Intent
+
+Vector has embraced the idea of **Paths**, which are analogous to CSS Selectors or `jq` selectors, that can represent an arbitrary selection of value(s). It is irrelevant to the user how the `LogEvent` is internally structured, so long as they can utilize these paths to work with a `LogEvent`.
+
+The `LogEvent` variant used under a two primary intents:
+
+- **Programmatically, through our internal code.** When used in this way, the `LogEvent` is being interacted with through static, safe interfaces and we can utilize the full type system available to us.
+- **Through configuration or FFI via knobs or scripts.** When used this way, the configuration format (at this time, TOML) may impact how aesthetic or expressive our configuration is. For example, TOML values like `a.b.c` as defined in a configuration. This is TOML table syntax, and we can't just magically treat it like a `Path`, a conversion step is necessary.
+
+## Understanding `LogEvent` in Vector 0.9.0
+
+Before we explore how to improve `LogEvent`, we must make sure we're familiar with the current one! Let's review the structure and some examples of lifetimes of this structure.
+
+### `LogEvent` Internal Structure
+
+It's current implementation meets our needs currently from several ergonomics angles (notably the flattening code), but it we must twist it in several painful ways to work with it (notably buffers to flatmaps to `BTree` style APIs).
```

Comment:
"it we must"? Also the leading "It's" should be "Its" and is somewhat dangling.

Reply:
Gosh you're right that's a terrible sentence.